### PR TITLE
[Backport 1.21] perf: add config to enable looking up local interface name for upstream connections (#19758)

### PR DESCRIPTION
**Pull request by pradeepcrao**
Commit Message:
Additional Description:
Backport #19758 to 1.21
Risk Level:
Testing: //test/...
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: test/mocks/upstream/cluster_info.h [hunk 1]
```diff
@@ -147,6 +147,7 @@ class MockClusterInfo : public ClusterInfo {
   MOCK_METHOD(bool, drainConnectionsOnHostRemoval, (), (const));
   MOCK_METHOD(bool, connectionPoolPerDownstreamConnection, (), (const));
   MOCK_METHOD(bool, warmHosts, (), (const));
+  MOCK_METHOD(bool, setLocalInterfaceNameOnUpstreamConnections, (), (const));
```
Comment (htuch) [inline test/mocks/upstream/cluster_info.h:150]: This looks almost identical to the original PR, so I can rubber stamp the backport. But, the tests in protocol_integration_test aren't here. Can you comment on these and any other delta? Thanks.

Comment (pradeepcrao) [reply test/mocks/upstream/cluster_info.h:150]: Not sure how I missed that. One of the test points actually fails in my workspace. Investigating it .

**Comment by phlax** [comment]
@pradeepcrao @htuch afaict this is the only blocker to releasing 1.21.2 can we land ?

**Comment by pradeepcrao** [comment]
> @pradeepcrao @htuch afaict this is the only blocker to releasing 1.21.2 can we land ?

Hi, I'm still blocked by a failing test and reached out to @junr03 (the author of the original PR) for help. IMHO, we can delay this backport to the next security release.

**Comment by phlax** [comment]
> Hi, I'm still blocked by a failing test 

CI seems to be passing - is that a downstream/external/local test?

> we can delay this backport to the next security release.

cool, i have prepared a release for the v1.20 branch today - i can look at the 1.21 branch tomorrow

File: test/mocks/upstream/cluster_info.h (hunk 1 above)
Comment (wrowe) [reply test/mocks/upstream/cluster_info.h:150]: Any update on this @pradeepcrao ?

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@pradeepcrao can you merge update the branch?

**Pull request merged**